cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamLeaderManager.java

```diff
@@ -0,0 +1,195 @@
+          streamsResourceCoordinatorClient.submitRequirement(builder.build()).get();
+        } catch (Throwable e) {
+          LOG.error("Could not create requirement for coordinator in Stream handler leader", e);
+          Throwables.propagate(e);
+        }
+      }
+
+      @Override
+      public void follower() {
+        LOG.info("Became Stream handler follower.");
+        if (streamsResourceCoordinator != null) {
+          streamsResourceCoordinator.stopAndWait();
+        }
+      }
+    });
+
+    handlerSubscription = streamsResourceCoordinatorClient.subscribe(handlerDiscoverable.getName(),
+                                                                     new StreamsLeaderHandler());
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    // revoke subscription to coordinator
+    if (leaderElection != null) {
+      leaderElection.stopAndWait();
+    }
+
+    if (handlerSubscription != null) {
+      handlerSubscription.cancel();
+    }
+
+    if (streamsResourceCoordinatorClient != null) {
+      streamsResourceCoordinatorClient.stopAndWait();
+    }
+  }
+
+  @Override
+  public void setHandlerDiscoverable(Discoverable discoverable) {
+    handlerDiscoverable = discoverable;
+  }
+
+  @Override
+  public ListenableFuture<Void> affectLeader(final String streamName) {
+    // modify the requirement to add the new stream as a new partition of the existing requirement
+    ListenableFuture<ResourceRequirement> future = streamsResourceCoordinatorClient.modifyRequirement(
+      STREAMS_RESOURCE, new ResourceModifier() {
+        @Nullable
+        @Override
+        public ResourceRequirement apply(@Nullable ResourceRequirement existingRequirement) {
+          Set<ResourceRequirement.Partition> partitions = existingRequirement.getPartitions();
+          ResourceRequirement.Partition newParition = new ResourceRequirement.Partition(streamName, 1);
+          if (partitions.contains(newParition)) {
+            return null;
+          }
+
+          ResourceRequirement.Builder builder = ResourceRequirement.builder(existingRequirement.getName());
+          builder.addPartition(newParition);
+          for (ResourceRequirement.Partition partition : partitions) {
+            builder.addPartition(partition);
+          }
+          return builder.build();
+        }
+      });
+    return Futures.transform(future, Functions.<Void>constant(null));
+  }
+
+  /**
+   * Class that defines the bahavior of a leader of a collection of Streams.
+   */
+  private final class StreamsLeaderHandler extends ResourceHandler {
+
+    protected StreamsLeaderHandler() {
+      super(handlerDiscoverable);
+    }
+
+    @Override
+    public void onChange(Collection<PartitionReplica> partitionReplicas) {
+      // TODO use the names of the ParitionReplicas to retrieve the streams names and do something on them
+      // TODO here, we are in the master for all the streams represented by the partitions replicas. We need to do
```

Comment:
here; we are in the master for all the streams represented by the partitions replicas.